### ISSUE TITLE
Terms and conditions page fails to load for users with lots of attributes

### ISSUE THREAD

**Issue by jaynis** [keycloakify/keycloakify#513]
Hi. I am experiencing a strange problem with the current version of the `keycloakify-starter` project when enabling the _Terms and Conditions_ action on a user which has lots (around hundred) of attributes. In such case the _Terms and Conditions_ page provided by `keycloakify-starter` loads forever until the request eventually times out in my proxy.

What I did is:

1. Clone the `keycloakify-starter` project into some dir, cd into it and run `yarn install && yarn build-keycloak-theme`
2. Add the resulting `keycloakify-starter-keycloak-theme-6.0.6.jar` to a `Dockerfile` which inherits from the Keycloak base image like so:
    ```Dockerfile
    FROM bitnami/keycloak:23.0.7-debian-12-r1
    
    ADD keycloakify-starter-keycloak-theme-6.0.6.jar /opt/bitnami/keycloak/providers/
    ```
3. Build the container and bring it into operation
4. Create a realm (e.g. `KeycloakifyRealm`) and a client (e.g. `keycloakify-client`), enable the `keycloakify-starter` theme on the realm, create a user, add lots of attributes to it and assign the _Terms and Conditions_ action to that user.
5. Try to login with some app using aforementioned client. The login theme itself is correctly working and displayed at this point.
6. Login with the user. The login hangs while Keycloak tries to load the following URL: `https://auth.example.com/realms/KeycloakifyRealm/login-actions/required-action?execution=TERMS_AND_CONDITIONS&client_id=keycloakify-client&tab_id=niIk2pg_7EY` (After a minute the upstream request to Keycloak times out, but I presume it would load forever).

As already mentioned, the custom theme itself works and if I disable the _Terms and Conditions_ action, login works flawlessly. Login with users with no or few attributes work as well, even if _Terms and Conditions_ is enabled. If I disable the custom theme and login using the Keycloak default theme, login works in any case. Summed up the behaviour I am experiencing is like follows:
| Theme       | Number of Attributes | Terms and Conditions enabled | Successful Login |
|-------------|----------------------|------------------------------|------------------|
| keycloakify-starter | 100                  | Yes                          | **No**               |
| keycloakify-starter | 0                    | Yes                          | **Yes**              |
| keycloakify-starter | 100                  | No                           | **Yes**              |
| keycloakify-starter | 0                    | No                           | **Yes**              |
| default     | 100                  | Yes                          | **Yes**              |
| default     | 0                    | Yes                          | **Yes**              |
| default     | 100                  | No                           | **Yes**              |
| default     | 0                    | No                           | **Yes**              |

It is indeed very strange and as of now I havent found out yet if it is really related to a large amount of attributes or to a specific attribute which I happen to have. All I can say is that with no attributes at all it works.

What I noticed as well is that during a login attempt the CPU usage of Keycloak immediately spikes for some time. If I do multiple login attempts simultaneously (which all hang) that could even result in Keycloak becoming completely unresponsive. While observing all this I cant see anything relevant in the Keycloak logs. Even with debug logs enabled there is nothing suspicious to see.

**Comment by garronej** [comment]
Hello @jaynis,

Thank you very much for reporting this issue; it has been fixed!

I wanted to mention that there is a command available for testing the theme locally in a Keycloak container.

![image](https://github.com/keycloakify/keycloakify/assets/6702424/a4304c8c-b333-48d8-bc1b-661e0936e731)

With just a single command, you can have a Keycloak instance ready to use with your theme pre-loaded. Plus, it supports hot reloading. This means you can manually modify the `.ftl` files, hit refresh, and see the changes immediately. Running `yarn build-keycloak-theme` again will also update the loaded theme.

This significantly simplifies the debugging and issue investigation process, making it 10 times easier.

In this particular case, the issue was a circular reference in the `client` FTL context. Although Keycloakify does not delve more than 10 levels deep into the tree, the complexity in this instance was `O(10^number of attributes)`, leading to very long load times rather than infinite loading.

I have implemented a special case to stop the recursion.

It's also nice to know that Keycloak makes user attributes, usernames, and other information available in the `terms.ftl` page. This feature enables the display of different terms based on the options selected by the user during registration.

**Comment by jaynis** [comment]
Thank you for the prompt fix @garronej!  I can confirm it is working now.

Thanks for pointing out the local deployment method, but with this issue where some users were working and others not for no apparent reason I got so paranoid at some point that I decided to test this in my real environment to rule out a bunch of other things :smile: . It took a while until I finally figured out it was related to the number of attributes.
